test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs

```diff
@@ -238,6 +242,47 @@ forceSeqToWhnf (SI.Seq SI.EmptyT)          k = k (SI.Seq SI.EmptyT)
 forceSeqToWhnf (SI.Seq (SI.Single a))      k = k (SI.Seq (SI.Single a))
 forceSeqToWhnf (SI.Seq (SI.Deep n l ft r)) k = k (SI.Seq (SI.Deep n l ft r))
 
+{-------------------------------------------------------------------------------
+  AllowThunksIn
+-------------------------------------------------------------------------------}
+
+data Record = Record
+  { field1 :: Int
+  , field2 :: Int
+  } deriving (Generic, Show)
+
+instance FromModel (AllowThunksIn '["field1"] Record) where
+  data Model (AllowThunksIn '["field1"] Record)
+    = RecordThunk   (Model (AllowThunksIn '["field1"] Record))
+    | RecordDefined (Model Int) (Model Int)
+
+  modelIsNF ctxt = \case
+    RecordThunk _ -> NotWHNF ctxt'
+    RecordDefined a b -> constrNF [modelIsNF ctxt' a, modelIsNF ctxt' b]
+    where
+      ctxt' = "Record" : ctxt
+
+  modelUnexpected ctxt m =
+      case m of
+        RecordThunk   _   -> Just ctxt'
+        RecordDefined _ y -> modelIsNF' ctxt' y
```

Comment:
Why the need to override this? Would the `modelIsNF` function on `Int` not suffice?

Reply:
I'm not sure what you're asking. The `modelIsNF'` function here is being called at `Int` here. Are you suggesting we use the default implementation of `modelUnexpected` in terms of `modelIsNF` `@Record`?

My interpretation of the intent behind `modelIsNF` is that it should answer whether the value is in NF or WHNF or not, regardless of whether the deviation from that is allowed. So the implementation here detects the permitted thunk in the first field.

The default implementation of `modelUnexpected` would produce a message for the thunk I wanted to ignore. To make the it do what we want, `modelIsNF` would need to ignore the permitted thunk. Is that fine?

Reply:
I guess what I'm asking is: why not just say

```
  modelUnexpected ctxt m =
      case m of
        RecordThunk   _   -> Just ctxt'
        RecordDefined _ y -> modelUnexpected ctxt y
      where
        ctxt' = "Record" : ctxt
```

saying: "If the record is defined, it has unexpected thunks only if the second field does". Now there is no reason to call that `modelIsNF'` function (or indeed abstract it out at all).

Reply:
Ah! I like it.